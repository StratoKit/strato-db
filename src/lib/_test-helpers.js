import DB from '../DB'
import ESDB from '../EventSourcingDB'
import EQ from '../EventQueue'
import JsonModel from '../JsonModel'

export {DB, JsonModel}

export const getModel = options => {
	const db = new DB()
	return db.addModel(JsonModel, {
		name: 'testing',
		keepRowId: false,
		...options,
	})
}

export const sharedSetup = getPromise => fn => {
	let promise
	return async () => {
		if (!promise) {
			promise = getPromise()
		}
		return fn(await promise)
	}
}

export const testModels = {
	count: {
		// shortName: 'c',
		columns: {
			total: {type: 'INTEGER'},
		},
		migrations: {
			init: {
				up({db, model, queue}) {
					// eslint-disable-next-line jest/no-standalone-expect
					expect(db).toBeTruthy()
					// eslint-disable-next-line jest/no-standalone-expect
					expect(queue).toBeTruthy()
					return model.set({id: 'count', total: 0, byType: {}})
				},
			},
		},
		preprocessor: async ({event}) => {
			if (event.type === 'error_pre') throw new Error('pre error for you')
		},
		reducer: async ({model, event: {type}}) => {
			if (type === 'error_reduce') throw new Error('error for you')
			if (!model.get) return false
			const c = (await model.get('count')) || {
				id: 'count',
				total: 0,
				byType: {},
			}
			c.total++
			c.byType[type] = (c.byType[type] || 0) + 1
			return {
				set: [c],
				// audit: '',
			}
		},
		deriver: async ({event}) => {
			if (event.type === 'error_derive') throw new Error('post error for you')
		},
	},
	ignorer: {
		// eslint-disable-next-line no-unused-vars
		reducer: args => {},
	},
	deriver: {
		deriver: async ({model, store, result, event}) => {
			if (result !== event.result[model.name]) {
				throw new Error('Expecting event.result as separate input')
			}
			if (event.result.count) {
				const currentCount = await store.count.get('count')
				await model.set({
					id: 'descCount',
					desc: `Total: ${currentCount.total}, seen types: ${Object.keys(
						currentCount.byType
					)}`,
				})
			}
		},
	},
}

const withDBs = async fn => {
	const db = new DB({name: 'D'})
	const queue = new EQ({
		db: new DB({name: 'Q'}),
		columns: {events: {type: 'JSON'}},
	})
	const ret = await fn(db, queue)
	await Promise.all([db.close(), queue.db.close()])
	return ret
}

/**
 * @param {Record<string,any>|function} modelsOrFn
 * @param {function} [fn]
 */
export const withESDB = (modelsOrFn, fn) => {
	let models
	if (typeof modelsOrFn === 'function') {
		// eslint-disable-next-line no-throw-literal
		if (fn) throw 'Use either .withESDB(fn) or .withESDB(models, fn)'
		fn = modelsOrFn
	} else {
		models = modelsOrFn
	}
	if (!models) models = testModels
	if (!fn) throw new Error('no fn passed to withESDB')
	return withDBs(async (db, queue) => {
		const eSDB = new ESDB({queue, models, name: 'E'})
		const out = await fn(eSDB, queue)
		await eSDB.close()
		return out
	})
}
