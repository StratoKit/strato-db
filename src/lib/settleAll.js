import {Sema} from 'async-sema'

// Only throw after all items are processed
/**
 * @template Item
 * @param {Item[]} items
 * @param {(item: Item) => Promise<void> | void} fn Function to call on each
 *   item.
 * @param {number} [maxConcurrent] Maximum functions running in parallel.
 */
export const settleAll = async (items, fn, maxConcurrent) => {
	let err, cb
	if (maxConcurrent) {
		const sema = new Sema(maxConcurrent)
		cb = async item => {
			await sema.acquire()
			try {
				return await fn(item)
			} finally {
				sema.release()
			}
		}
	} else {
		cb = fn
	}
	await Promise.all(
		items.map(async i => {
			try {
				await cb(i)
			} catch (error) {
				// last one wins
				err = error
			}
		})
	)
	if (err) throw err
}
