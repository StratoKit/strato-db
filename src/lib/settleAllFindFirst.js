export /**
 *	Execute all fn (asynchronously), if fn returns values, it returns the first result
 *
 * @param {[objects]} items
 * @param {function} fn
 * @returns {any} first positive result
 */
const settleAllFindFirst = async (items, fn) => {
	let err
	const results = await Promise.all(
		items.map(async i => {
			try {
				const res = await fn(i)
				if (res) return res
			} catch (error) {
				// last one wins
				err = error
			}
		})
	)
	if (err) throw err
	if (results?.filter(Boolean).length) return results[0]
	return false
}
