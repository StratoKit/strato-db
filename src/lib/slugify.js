import {deburr} from 'lodash'

const abc = 'abcdefghijklmnopqrstuvwxyz0123456789'
export const randomString = (/** @type {number} */ n) =>
	// eslint-disable-next-line unicorn/no-new-array
	Array.apply(null, new Array(n))
		.map(() => {
			return abc.charAt(Math.floor(Math.random() * abc.length))
		})
		.join('')

export const slugifyString = (name, alwaysResult) => {
	// extract name from i18n objects
	const t =
		typeof name === 'string'
			? name
			: typeof name === 'number'
			? name.toString()
			: name && typeof name === 'object'
			? Object.values(name).find(v => typeof v === 'string' && v)
			: null
	if (!t) {
		if (alwaysResult) return randomString(12)
		throw new Error(`Cannot slugify ${name}`)
	}
	return encodeURIComponent(deburr(t).trim())
		.replaceAll(/(%..|['()_~])/g, '-')
		.replaceAll(/--+/g, '-')
		.toLowerCase()
		.replaceAll(/(^[^\da-z]+|[^\da-z]+$)/g, '')
		.slice(0, 30)
}

// This is not race-safe - only use for write-seldomn things like backoffice or inside transactions
export const uniqueSlugId = async (model, name, colName, currentId) => {
	const slug = slugifyString(name, true)
	let id = slug
	let i = 1
	const where = currentId && {
		[`${model.idColQ} IS NOT ?`]: [currentId],
	}

	while (await model.exists({[colName]: id}, {where})) {
		id = `${slug}-${++i}`
	}
	return id
}
