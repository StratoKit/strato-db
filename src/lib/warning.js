export const DEV = process.env.NODE_ENV !== 'production'
export let deprecated, unknown

if (DEV) {
	const warned = {}
	const warner = type => (tag, msg, conditionFn) => {
		if (warned[tag]) return
		if (conditionFn && !conditionFn()) return
		warned[tag] = true
		// eslint-disable-next-line no-console
		console.warn(new Error(`!!! ${type} ${msg}`))
	}
	deprecated = warner('DEPRECATED')
	unknown = warner('UNKNOWN')
} else {
	deprecated = () => {}
	unknown = () => {}
}
