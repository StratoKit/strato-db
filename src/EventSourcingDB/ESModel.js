// Drop-in replacement for JsonModel
// Caveats:
// * `.update()` returns the current object at the time of returning, not the one that was updated
//
// Events all type `es/name` and data `[actionEnum, id, obj, meta]`
// The id is assigned by the preprocessor except for RM

import JsonModel from '../JsonModel'
import {DEV} from '../lib/warning'
import {isEqual} from 'lodash'
import applyResult from './applyResult'

export const undefToNull = data => {
	if (data == null) return null
	if (typeof data !== 'object') return data
	if (Array.isArray(data)) return data.map(element => undefToNull(element))
	if (Object.getPrototypeOf(data) !== Object.prototype) return data
	const out = {}
	for (const [key, value] of Object.entries(data)) {
		out[key] = undefToNull(value)
	}
	return out
}

export const getId = async (model, data) => {
	let id = data[model.idCol]
	if (id == null) {
		// Be sure to call with model as this, like in JsonModel
		id = await model.columns[model.idCol].value.call(model, data)
	}
	// This can only happen for integer ids
	if (id == null) id = await model.getNextId()
	return id
}

// Calculate the update given two objects that went
// through JSON stringify+parse
const calcUpd = (idCol, prev, obj, complete) => {
	const out = {}
	let changed = false
	for (const [key, value] of Object.entries(obj)) {
		const pVal = prev[key]
		if (value == null && pVal != null) {
			out[key] = null
			changed = true
		} else if (!isEqual(value, pVal)) {
			out[key] = value
			changed = true
		}
	}
	if (complete)
		for (const key of Object.keys(prev))
			if (!(key in obj)) {
				out[key] = null
				changed = true
			}
	if (changed) {
		out[idCol] = prev[idCol]
		return out
	}
	return undefined
}

/**
 * ESModel is a drop-in wrapper around JsonModel to turn changes into events.
 *
 * Use it to convert your database to be event sourcing
 *
 * Event data is encoded as an array: `[subtype, id, data, meta]`
 * Subtype is one of `ESModel.(REMOVE|SET|INSERT|UPDATE|SAVE)`.
 * `id` is filled in by the preprocessor at the time of the event.
 * `meta` is free-form data about the event. It is just stored in the history
 * table.
 *
 * For example: `model.set({foo: true})` would result in the event `[1, 1, {foo:
 * true}]`
 *
 * @augments JsonModel
 */
class ESModel extends JsonModel {
	static REMOVE = 0
	static SET = 1
	static INSERT = 2
	static UPDATE = 3
	static SAVE = 4

	/**
	 * Creates a new ESModel model, called by DB.
	 *
	 * @class
	 * @param {function} dispatch      - the {@link ESDB} dispatch function.
	 * @param {boolean}  [init]        - emit an event with type
	 *                                 `es/INIT:${modelname}` at table creation
	 *                                 time, to be used by custom reducers.
	 * @param {Object}   [...options]  - other params are passed to JsonModel.
	 */
	constructor({dispatch, init, emitter, ...options}) {
		super({
			...options,
			migrations: {
				...options.migrations,
				'0_init':
					init &&
					(({queue}) => {
						// Don't wait for add Promise to prevent deadlock
						queue.add(this.INIT)
					}),
			},
		})
		this.dispatch = dispatch
		this.writable = false
		const clearMax = () => {
			this._maxId = 0
		}
		options.db.on('begin', clearMax)
		emitter.on('result', clearMax)
		emitter.on('error', clearMax)
	}

	TYPE = `es/${this.name}`

	INIT = `es/INIT:${this.name}`

	/**
	 * Slight hack: use the writable state to fall back to JsonModel behavior.
	 * This makes deriver and migrations work without changes.
	 * Note: while writable, no events are created. Be careful.
	 *
	 * @param {boolean} state  - writeable or not.
	 */
	setWritable(state) {
		this.writable = state
	}

	/**
	 * Insert or replace the given object into the database.
	 *
	 * @param {Object}  obj           - the object to store. If there is no `id`
	 *                                value (or whatever the `id` column is named),
	 *                                one is assigned automatically.
	 * @param {boolean} [insertOnly]  - don't allow replacing existing objects.
	 * @param {boolean} [noReturn]    - do not return the stored object; an
	 *                                optimization.
	 * @param {*}       [meta]        - extra metadata to store in the event but
	 *                                not in the object.
	 * @returns {Promise<Object>} - if `noReturn` is false, the stored object is
	 *                            fetched from the DB.
	 */
	async set(obj, insertOnly, noReturn, meta) {
		if (DEV && noReturn != null && typeof noReturn !== 'boolean')
			throw new Error(`${this.name}: meta argument is now in fourth position`)
		if (this.writable) {
			const id = obj[this.idCol]
			if (this._maxId && id > this._maxId) this._maxId = id
			return super.set(obj, insertOnly, noReturn)
		}

		const d = [insertOnly ? ESModel.INSERT : ESModel.SET, null, obj]
		if (meta) d[3] = meta

		const {data, result} = await this.dispatch(this.TYPE, d)
		const id = data[1]

		const r = result[this.name]
		if (r && r.esFail) throw new Error(`${this.name}.set ${id}: ${r.esFail}`)

		// We have to get because we don't know what calculated values did
		// Unfortunately, this might be the object after a later event
		return noReturn ? undefined : this.get(id)
	}

	/**
	 * Update an existing object.
	 *
	 * @param {Object}  o           - the data to store.
	 * @param {boolean} [upsert]    - if `true`, allow inserting if the object
	 *                              doesn't exist.
	 * @param {boolean} [noReturn]  - do not return the stored object; an
	 *                              optimization.
	 * @param {*}       [meta]      - extra metadata to store in the event at
	 *                              `data[3]` but not in the object.
	 * @returns {Promise<Object>} - if `noReturn` is false, the stored object is
	 *                            fetched from the DB.
	 */
	async update(o, upsert, noReturn, meta) {
		if (DEV && noReturn != null && typeof noReturn !== 'boolean')
			throw new Error(`${this.name}: meta argument is now in fourth position`)

		if (this.writable) return super.update(o, upsert, noReturn)

		if (DEV && noReturn != null && typeof noReturn !== 'boolean')
			throw new Error(`${this.name}: meta argument is now in fourth position`)
		let id = o[this.idCol]
		if (id == null && !upsert) throw new TypeError('No ID specified')

		const d = [upsert ? ESModel.SAVE : ESModel.UPDATE, null, undefToNull(o)]
		if (meta) d.push(meta)

		const {data, result} = await this.dispatch(this.TYPE, d)
		id = data[1]

		const r = result[this.name]
		if (r && r.esFail) throw new Error(`${this.name}.update ${id}: ${r.esFail}`)

		// We have to get because we don't know what calculated values did
		// Unfortunately, this might be the object after a later event
		return this.get(id)
	}

	updateNoTrans(obj, upsert) {
		if (this.writable) return super.updateNoTrans(obj, upsert)
		throw new Error('Non-transactional changes are not possible with ESModel')
	}

	/**
	 * Remove an object.
	 *
	 * @param {Object | string | integer} idOrObj
	 * - the id or the object itself.
	 * @param {*} meta
	 * - metadata, attached to the event only, at `data[3]`
	 * @returns {Promise<boolean>} - always returns true.
	 */
	async remove(idOrObj, meta) {
		if (this.writable) return super.remove(idOrObj)
		const id = typeof idOrObj === 'object' ? idOrObj[this.idCol] : idOrObj
		if (id == null) throw new TypeError('No ID specified')

		const d = [ESModel.REMOVE, id]
		if (meta) d[3] = meta

		await this.dispatch(this.TYPE, d)
		return true
	}

	/** changeId: not implemented yet, had no need so far */
	changeId(oldId, newId) {
		if (this.writable) return super.changeId(oldId, newId)
		throw new Error(`ESModel doesn't support changeId yet`)
	}

	_maxId = 0
	_maxIdP = null
	_lastUV = 0

	/**
	 * Returns the next available integer ID for the model.
	 * Calling this multiple times during a redux cycle will give increasing
	 * numbers even though the database table doesn't change.
	 * Use this from the redux functions to assign unique ids to new objects.
	 *
	 * @returns {Promise<number>} - the next usable ID.
	 */
	async getNextId() {
		if (!this._maxId) {
			if (!this._maxIdP)
				this._maxIdP = this.max(this.idCol).then(m => {
					this._maxId = m
					return m
				})
			await this._maxIdP
			this._maxIdP = null
		}
		return ++this._maxId
	}

	/**
	 * Applies the result from the reducer.
	 *
	 * @param {Object} result  - free-form change descriptor.
	 * @returns {Promise<void>} - Promise for completion.
	 */
	async applyResult(result) {
		if (result.esFail) return
		return applyResult(this, {...result, esFail: undefined})
	}

	/**
	 * Assigns the object id to the event at the start of the cycle.
	 * When subclassing ESModel, be sure to call this too (`ESModel.preprocessor(arg)`)
	 */
	static async preprocessor({model, event, isMainEvent}) {
		if (isMainEvent) this._maxId = 0
		if (event.type !== model.TYPE) return
		if (event.data[0] > ESModel.REMOVE) {
			// Always overwrite, so repeat events get correct ids
			// eslint-disable-next-line require-atomic-updates
			event.data[1] = await getId(model, event.data[2])
			return event
		}
	}

	/**
	 * Calculates the desired change ESModel will only emit `rm`, `ins`, `upd` and
	 * `esFail`
	 *
	 * @param {Object} model  - the model.
	 * @param {Event}  event  - the event.
	 * @returns {Promise<Object>} - the result object in the format JsonModel
	 *                            likes.
	 */
	static async reducer({model, event: {type, data}}) {
		if (!model || type !== model.TYPE) return false

		let [action, id, obj] = data
		if (action === ESModel.REMOVE) {
			if (await model.exists({[model.idCol]: id})) return {rm: [id]}
			return false
		}

		if (obj[model.idCol] == null) obj = {...obj, [model.idCol]: id}

		const prev = await model.get(id)
		let update
		if (prev) {
			if (action === ESModel.INSERT) return {esFail: 'EEXIST'}
			update = calcUpd(model.idCol, prev, obj, action === ESModel.SET)
			return update ? {upd: [update]} : false
		}
		if (action === ESModel.UPDATE) return {esFail: 'ENOENT'}
		return {ins: [obj]}
	}
}

export default ESModel
