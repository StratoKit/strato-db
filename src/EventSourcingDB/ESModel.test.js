import ESModel, {undefToNull, getId} from './ESModel'
import {withESDB} from '../lib/_test-helpers'

class ESModelCustomId extends ESModel {
	constructor(options) {
		super({
			...options,
			columns: {
				myId: {
					index: true,
				},
			},
			idCol: 'myId',
		})
	}
}

class ESModelIntId extends ESModel {
	constructor(options) {
		super({
			...options,
			columns: {
				myId: {
					index: true,
					type: 'INTEGER',
				},
				calc: {value: o => o.myId || 0},
			},
			idCol: 'myId',
		})
	}
}

const sampleObject = {id: 'asd', meep: 'moop', top: 'kek'}

describe('undefToNull', () => {
	test('works', () => {
		const obj = {
			a: 'b',
			c: 4,
			d: new Date('2007-01-01'),
			nested: {
				stuff: 'test',
				asd: undefined,
				dsa: null,
			},
			array: [
				'two plus two is',
				undefined,
				{minus: 1, thats: '?'},
				null,
				'quick maths',
			],
			zxc: null,
			cxz: undefined,
		}
		expect(undefToNull(obj)).toEqual({
			...obj,
			nested: {...obj.nested, asd: null},
			array: [
				'two plus two is',
				null,
				{minus: 1, thats: '?'},
				null,
				'quick maths',
			],
			cxz: null,
		})
	})
})

describe('getId', () => {
	test('standard settings', () =>
		withESDB({test: {Model: ESModel}}, async eSDB => {
			expect(await getId(eSDB.store.test, {top: 'kek'})).toBeTruthy()
		}))

	test('custom idCol', () =>
		withESDB({test: {Model: ESModelCustomId}}, async eSDB => {
			expect(await getId(eSDB.store.test, {top: 'kek'})).toBeTruthy()
		}))

	test('integer idCol', () =>
		withESDB({test: {Model: ESModelIntId}}, async eSDB => {
			expect(await getId(eSDB.store.test, {top: 'kek'})).toBe(1)
			expect(await getId(eSDB.store.test, {top: 'kek'})).toBe(2)
			await eSDB.store.test.set({myId: 7, asd: 'dsa'})
			expect(await getId(eSDB.store.test, {moop: 'meep'})).toBe(8)
		}))
})

describe('ESModel', () => {
	test('create', () =>
		withESDB({test: {Model: ESModel}}, async eSDB => {
			const result = await eSDB.store.test.all()
			expect(result).toEqual([])
		}))

	test('init', () =>
		withESDB(
			{
				m: {
					init: true,
					reducer: ({model, event}) =>
						event.type === model.INIT ? {ins: [{id: 'yey'}]} : false,
				},
			},
			async eSDB => {
				await eSDB.waitForQueue()
				const {m} = eSDB.store
				expect(await m.exists({id: 'yey'})).toBeTruthy()
			}
		))

	describe('.set()', () => {
		test('with id', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				expect(await eSDB.store.test.set(sampleObject)).toEqual(sampleObject)
			}))

		test('without id', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				const newObject = await eSDB.store.test.set({
					...sampleObject,
					id: undefined,
				})
				expect(await eSDB.store.test.get(newObject.id)).toEqual(newObject)
			}))

		test('without integer id', () =>
			withESDB({test: {Model: ESModelIntId}}, async eSDB => {
				const sampleWithoutId = {
					...sampleObject,
					id: undefined,
				}
				await Promise.all([
					eSDB.store.test.set(sampleWithoutId),
					eSDB.store.test.set(sampleWithoutId),
					eSDB.store.test.set(sampleWithoutId),
				])
				expect(await getId(eSDB.store.test, {top: 'kek'})).toEqual(4)
				expect(await eSDB.store.test.count()).toEqual(3)
			}))

		test('calculated value', () =>
			withESDB({test: {Model: ESModelIntId}}, async eSDB => {
				const sampleWithoutId = {
					...sampleObject,
					id: undefined,
				}
				await expect(
					eSDB.store.test.set(sampleWithoutId)
				).resolves.toHaveProperty('calc', 1)
				await expect(
					eSDB.store.test.update({myId: 1, calc: 5})
				).resolves.toHaveProperty('calc', 1)
			}))

		test('insertOnly', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await expect(
					Promise.all([
						eSDB.store.test.set(sampleObject, true),
						eSDB.store.test.set(sampleObject, true),
					])
				).rejects.toThrow('EEXIST')
			}))
	})

	describe('.update()', () => {
		test('works', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await eSDB.store.test.set(sampleObject)
				const newObject = {
					id: sampleObject.id,
					top: 'bottom',
					new: 'prop',
				}
				await eSDB.store.test.update(newObject)
				const items = await eSDB.store.test.all()
				expect(items).toHaveLength(1)
				expect(items[0]).toEqual({...sampleObject, ...newObject})
			}))

		test('without id', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await expect(eSDB.store.test.update({top: 'kek'})).rejects.toThrow(
					'No ID specified'
				)
			}))

		test('undefined values', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await eSDB.store.test.set(sampleObject)
				expect(
					await eSDB.store.test.update({id: sampleObject.id, top: undefined})
				).toEqual({...sampleObject, top: undefined})
			}))

		test('non-existent object', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await expect(eSDB.store.test.update(sampleObject)).rejects.toThrow(
					'ENOENT'
				)
			}))

		test('upsert', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				expect(await eSDB.store.test.update(sampleObject, true)).toEqual(
					sampleObject
				)
			}))

		test('upsert without id', () =>
			withESDB({test: {Model: ESModelIntId}}, async eSDB => {
				expect(await eSDB.store.test.update({top: 'kek'}, true)).toEqual({
					myId: 1,
					top: 'kek',
					calc: 1,
				})
			}))
	})

	describe('.remove()', () => {
		test('remove by id', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await eSDB.store.test.set(sampleObject)
				expect(await eSDB.store.test.remove(sampleObject.id)).toEqual(true)
				expect(await eSDB.store.test.all()).toEqual([])
			}))

		test('remove by object', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await eSDB.store.test.set(sampleObject)
				expect(await eSDB.store.test.remove(sampleObject)).toEqual(true)
				expect(await eSDB.store.test.all()).toEqual([])
			}))

		test('remove by object without id', () =>
			withESDB({test: {Model: ESModel}}, async eSDB => {
				await eSDB.store.test.set(sampleObject)
				await expect(
					eSDB.store.test.remove({...sampleObject, id: undefined})
				).rejects.toThrow()
			}))
	})

	describe('preprocessor', () => {
		test('works', () => {
			let ok
			return withESDB(
				{
					m: {
						idCol: 'hi',
						reducer: args => {
							if (!args.model) return false
							if (
								args.event.type === args.model.TYPE &&
								args.event.data[1] === 'a'
							)
								ok = true
							return ESModel.reducer(args)
						},
					},
				},
				async eSDB => {
					const {m} = eSDB.store
					await m.set({hi: 'a', meep: 'moop'})
					expect(ok).toBe(true)
				}
			)
		})
	})

	describe('itemClass', () => {
		class Foo {
			getF() {
				return this.f
			}
		}

		test('set', () =>
			withESDB(
				{
					m: {
						Model: class extends ESModel {
							constructor(o) {
								super({...o, ItemClass: Foo})
							}
						},
					},
				},
				async eSDB => {
					const {m} = eSDB.store
					const setted = await m.set({id: 4, f: 'meep'})
					expect(setted instanceof Foo).toBe(true)
					expect(setted.getF()).toBe('meep')
				}
			))
	})

	describe('.getNextId()', () => {
		test('works', () =>
			withESDB({m: {columns: {id: {type: 'INTEGER'}}}}, async eSDB => {
				const {m} = eSDB.store
				await expect(m.getNextId()).resolves.toBe(1)
				await m.set({id: 1})
				await expect(m.getNextId()).resolves.toBe(2)
				await eSDB.dispatch(m.TYPE, [ESModel.INSERT, 5, {id: 5}])
				await expect(m.getNextId()).resolves.toBe(6)
			}))

		test('concurrent', () =>
			withESDB({m: {columns: {id: {type: 'INTEGER'}}}}, async eSDB => {
				const {m} = eSDB.store
				expect(await Promise.all([m.getNextId(), m.getNextId()])).toEqual([
					1, 2,
				])
			}))

		test('run in subevent', () =>
			withESDB(
				{
					test: {
						Model: class ESModelSubevents extends ESModel {
							constructor(options) {
								super({
									...options,
									columns: {
										id: {
											type: 'INTEGER',
										},
									},
								})
							}

							static async reducer(arg) {
								const {event, addEvent, model} = arg
								const result = (await super.reducer(arg)) || event.result || {}
								if (event.type === 'SUBEVENT') {
									return {
										ins: [{id: await model.getNextId(), fromSubevent: true}],
									}
								}
								if (
									result &&
									result.upd &&
									result.upd.some(e => e.triggerSub)
								) {
									addEvent('SUBEVENT')
								}
								return result
							}

							static async deriver({model, result}) {
								if (
									result &&
									result.upd &&
									result.upd.some(e => e.triggerSub)
								) {
									model.set({fromDeriver: true})
								}
							}
						},
					},
				},
				async eSDB => {
					await eSDB.store.test.set({name: 'id 1'})
					await eSDB.store.test.set({name: 'id 2'})
					await eSDB.store.test.set({name: 'id 3'})
					await eSDB.store.test.set({name: 'id 4'})
					await eSDB.store.test.set({name: 'id 5'})
					await eSDB.store.test.update({id: 1, triggerSub: 'oh yiss'})

					expect(await eSDB.store.test.get(6)).toHaveProperty(
						'fromDeriver',
						true
					)
					expect(await eSDB.store.test.get(7)).toHaveProperty(
						'fromSubevent',
						true
					)
				}
			))
	})

	describe('events', () => {
		test('work', () =>
			withESDB({m: {columns: {id: {type: 'INTEGER'}}}}, async (eSDB, queue) => {
				const {m} = eSDB.store
				await m.set({meep: 'moop'})
				await m.update({id: 1, beep: 'boop'})
				await m.set({meep: 'moop'}, true)
				await m.update({id: 3, beep: 'boop'}, true)
				await m.remove(3)
				const events = await queue.all()
				expect(
					events.map(e => {
						e.ts = 0
						return e
					})
				).toMatchSnapshot()
			}))

		test('are updated', () =>
			withESDB(
				{m: {idCol: 'v', columns: {v: {type: 'INTEGER'}}}},
				async (eSDB, queue) => {
					const {m} = eSDB.store
					expect(await m.set({meep: 'moop'})).toEqual({v: 1, meep: 'moop'})
					expect(await m.set({v: 1, meep: 'moop'})).toEqual({
						v: 1,
						meep: 'moop',
					})
					expect(await m.set({v: 1, beep: 'boop', a: [null, 3]})).toEqual({
						v: 1,
						beep: 'boop',
						a: [null, 3],
					})
					expect(await m.update({v: 1, beep: 'boop'})).toEqual({
						v: 1,
						beep: 'boop',
						a: [null, 3],
					})
					expect(await m.update({v: 1, beep: 'foop', a: [null, 3]})).toEqual({
						v: 1,
						beep: 'foop',
						a: [null, 3],
					})
					const events = await queue.all()
					expect(
						events.map(e => {
							e.ts = 0
							return e
						})
					).toMatchSnapshot()
				}
			))

		test('hold metadata', () =>
			withESDB({m: {columns: {id: {type: 'INTEGER'}}}}, async (eSDB, queue) => {
				const {m} = eSDB.store
				await m.set({meep: 'moop'}, null, true, {meta: 1})
				await m.update({id: 1, beep: 'boop'}, null, true, {meta: 2})
				await m.set({meep: 'moop'}, true, true, {meta: 3})
				await m.update({id: 3, beep: 'boop'}, true, true, 'hi')
				await m.set({})
				await m.remove(3, {meta: 4})
				await m.remove(2)
				const events = await queue.all()
				expect(events[0].data).toHaveLength(4)
				expect(events[4].data).toHaveLength(3)
				expect(events[6].data).toHaveLength(2)
				expect(events.map(e => e.data[3])).toEqual([
					{meta: 1},
					{meta: 2},
					{meta: 3},
					'hi',
					undefined,
					{meta: 4},
					undefined,
				])
			}))

		describe('event creators', () => {
			const fake = {
				registerMigrations: () => {
					// mock
				},
				on: () => {
					// mock
				},
				setMaxListeners: () => 10,
				getMaxListeners: () => 10,
			}
			const m = new ESModel({
				name: 'hi',
				db: fake,
				emitter: fake,
			})
			test.each([
				[
					'set',
					[{foo: true}],
					{type: 'es/hi', data: [ESModel.SET, null, {foo: true}]},
				],
				[
					'set',
					[{id: 'hi', bar: true}, true, {meta: 'hello'}],
					{
						type: 'es/hi',
						data: [
							ESModel.INSERT,
							null,
							{id: 'hi', bar: true},
							{meta: 'hello'},
						],
					},
				],
				[
					'update',
					[{id: 7, foo: true, bar: undefined}],
					{
						type: 'es/hi',
						data: [ESModel.UPDATE, null, {id: 7, foo: true, bar: null}],
					},
				],
				[
					'update', // with upsert
					[{foo: true}, true],
					{type: 'es/hi', data: [ESModel.SAVE, null, {foo: true}]},
				],
				[
					'remove',
					[{id: 'hi', f: 1}],
					{type: 'es/hi', data: [ESModel.REMOVE, 'hi']},
				],
				['remove', [7], {type: 'es/hi', data: [ESModel.REMOVE, 7]}],
			])('.event.%s%O => %O', (action, input, output) =>
				expect(m.event[action](...input)).toEqual(output)
			)
		})
	})
})
