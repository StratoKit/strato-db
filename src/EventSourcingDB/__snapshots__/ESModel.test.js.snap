// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESModel events are updated 1`] = `
Array [
  Object {
    "data": Array [
      1,
      1,
      Object {
        "meep": "moop",
      },
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "meep": "moop",
            "v": 1,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 1,
  },
  Object {
    "data": Array [
      1,
      1,
      Object {
        "meep": "moop",
        "v": 1,
      },
    ],
    "result": Object {},
    "ts": 0,
    "type": "es/m",
    "v": 2,
  },
  Object {
    "data": Array [
      1,
      1,
      Object {
        "a": Array [
          null,
          3,
        ],
        "beep": "boop",
        "v": 1,
      },
    ],
    "result": Object {
      "m": Object {
        "upd": Array [
          Object {
            "a": Array [
              null,
              3,
            ],
            "beep": "boop",
            "meep": null,
            "v": 1,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 3,
  },
  Object {
    "data": Array [
      3,
      1,
      Object {
        "beep": "boop",
        "v": 1,
      },
    ],
    "result": Object {},
    "ts": 0,
    "type": "es/m",
    "v": 4,
  },
  Object {
    "data": Array [
      3,
      1,
      Object {
        "a": Array [
          null,
          3,
        ],
        "beep": "foop",
        "v": 1,
      },
    ],
    "result": Object {
      "m": Object {
        "upd": Array [
          Object {
            "beep": "foop",
            "v": 1,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 5,
  },
]
`;

exports[`ESModel events work 1`] = `
Array [
  Object {
    "data": Array [
      1,
      1,
      Object {
        "meep": "moop",
      },
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "id": 1,
            "meep": "moop",
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 1,
  },
  Object {
    "data": Array [
      3,
      1,
      Object {
        "beep": "boop",
        "id": 1,
      },
    ],
    "result": Object {
      "m": Object {
        "upd": Array [
          Object {
            "beep": "boop",
            "id": 1,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 2,
  },
  Object {
    "data": Array [
      2,
      2,
      Object {
        "meep": "moop",
      },
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "id": 2,
            "meep": "moop",
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 3,
  },
  Object {
    "data": Array [
      4,
      3,
      Object {
        "beep": "boop",
        "id": 3,
      },
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "beep": "boop",
            "id": 3,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 4,
  },
  Object {
    "data": Array [
      0,
      3,
    ],
    "result": Object {
      "m": Object {
        "rm": Array [
          3,
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 5,
  },
]
`;
